src/wallet/wallet.cpp

```diff
@@ -857,80 +858,128 @@ bool CWallet::LoadUnifiedAddressMetadata(const ZcashdUnifiedAddressMetadata &add
+                                ufvk->Address(j.value().first, CWallet::DefaultReceiverTypes(nHeight)));
+                        return std::make_pair(PaymentAddress{defaultUA.first}, RecipientType::WalletExternalAddress);
+                    } else {
+                        return std::make_pair(PaymentAddress{addr}, RecipientType::WalletInternalAddress);
+                    }
                 }
-            },
-            [&](const OrchardRawAddress& addr) {
-                auto ua = self->FindUnifiedAddressByReceiver(addr);
-                if (ua.has_value()) {
-                    return libzcash::PaymentAddress{ua.value()};
-                } else if (ufvk.has_value() && ufvk->GetOrchardKey().has_value()) {
-                    auto orchardKey = ufvk->GetOrchardKey().value();
-                    auto j = orchardKey.ToIncomingViewingKey().DecryptDiversifier(addr);
-                    if (j.has_value()) {
-                        auto genResult = ufvk->Address(j.value(), CWallet::DefaultReceiverTypes());
+            }
+
+            // We don't produce internal change addresses for legacy Sapling
+            // addresses, so this must be a counterparty address
+            return std::make_pair(PaymentAddress{addr}, RecipientType::CounterpartyAddress);
+        },
+        [&](const OrchardRawAddress& addr) {
+            auto ua = self->FindUnifiedAddressByReceiver(addr);
+            if (ua.has_value()) {
+                // UAs are always external addresses
+                return std::make_pair(PaymentAddress{ua.value()}, RecipientType::WalletExternalAddress);
+            } else if (ufvk.has_value() && ufvk->GetOrchardKey().has_value()) {
+                auto orchardKey = ufvk->GetOrchardKey().value();
+                auto j = orchardKey.DecryptDiversifier(addr);
+                if (j.has_value()) {
+                    if (j.value().second) {
+                        // Attempt to reproduce the original unified address
+                        auto genResult = ufvk->Address(j.value().first, CWallet::DefaultReceiverTypes(nHeight));
                         auto defaultUA = std::get_if<std::pair<UnifiedAddress, diversifier_index_t>>(&genResult);
                         if (defaultUA != nullptr) {
-                            return libzcash::PaymentAddress{defaultUA->first};
+                            return std::make_pair(PaymentAddress{defaultUA->first}, RecipientType::WalletExternalAddress);
                         }
+                    } else {
+                        return std::make_pair(PaymentAddress{UnifiedAddress::ForSingleReceiver(addr)}, RecipientType::WalletInternalAddress);
                     }
                 }
-
-                return libzcash::PaymentAddress{UnifiedAddress::ForSingleReceiver(addr)};
             }
-        }, recipient);
-    };
+
+            return std::make_pair(PaymentAddress{UnifiedAddress::ForSingleReceiver(addr)}, RecipientType::CounterpartyAddress);
```

Comment:
Here we're fine, because we've never generated Orchard receivers other than as part of UAs (and we do not detect them independently from UAs).